src/spaces/jolt_physics_direct_space_state_3d.cpp

```diff
@@ -983,28 +1019,75 @@ bool JoltPhysicsDirectSpaceState3D::_body_motion_collide(
 		const JoltShapedObjectImpl3D* collider = collider_jolt_body.as_shaped();
 		ERR_FAIL_NULL_D(collider);
 
-		const Vector3 position = base_offset + to_godot(hit.mContactPointOn2);
-
 		const int32_t local_shape = p_body.find_shape_index(hit.mSubShapeID1);
 		ERR_FAIL_COND_D(local_shape == -1);
 
 		const int32_t collider_shape = collider->find_shape_index(hit.mSubShapeID2);
 		ERR_FAIL_COND_D(collider_shape == -1);
 
-		PhysicsServer3DExtensionMotionCollision& collision = p_result->collisions[count++];
-
-		collision.position = position;
-		collision.normal = to_godot(-hit.mPenetrationAxis.Normalized());
-		collision.collider_velocity = collider->get_velocity_at_position(position);
-		collision.collider_angular_velocity = collider->get_angular_velocity();
-		collision.depth = penetration_depth;
-		collision.local_shape = local_shape;
-		collision.collider_id = collider->get_instance_id();
-		collision.collider = collider->get_rid();
-		collision.collider_shape = collider_shape;
+		for (JPH::Vec3 contact_point : contact_points2) {
+			const Vector3 position = base_offset + to_godot(contact_point);
+
+			PhysicsServer3DExtensionMotionCollision& collision = p_result->collisions[count++];
+
+			collision.position = position;
+			collision.normal = to_godot(-hit.mPenetrationAxis.Normalized());
+			collision.collider_velocity = collider->get_velocity_at_position(position);
+			collision.collider_angular_velocity = collider->get_angular_velocity();
+			collision.depth = penetration_depth;
+			collision.local_shape = local_shape;
+			collision.collider_id = collider->get_instance_id();
+			collision.collider = collider->get_rid();
+			collision.collider_shape = collider_shape;
+
+			if (count == p_max_collisions) {
+				break;
+			}
+		}
+
+		if (count == p_max_collisions) {
+			break;
+		}
 	}
 
 	p_result->collision_count = count;
 
 	return count > 0;
 }
+
+void JoltPhysicsDirectSpaceState3D::_generate_manifold(
+	const JPH::CollideShapeResult& p_hit,
+	JPH::ContactPoints& p_contact_points1,
+	JPH::ContactPoints& p_contact_points2
+#ifdef JPH_DEBUG_RENDERER
+	,
+	JPH::RVec3Arg p_center_of_mass
+#endif // JPH_DEBUG_RENDERER
+) const {
+	const JPH::PhysicsSystem& physics_system = space->get_physics_system();
+	const JPH::PhysicsSettings& physics_settings = physics_system.GetPhysicsSettings();
+
+	JPH::ManifoldBetweenTwoFaces(
+		p_hit.mContactPointOn1,
+		p_hit.mContactPointOn2,
+		p_hit.mPenetrationAxis,
+		physics_settings.mManifoldToleranceSq,
```

Comment:
@jrouwe Forgive me for pinging you twice, but should I also be adding the squared speculative contact distance here, like you do in your non-reducing collector?

Reply:
No you should not add that here, the parameter is named a bit strangely maybe (also this function was never intended to be used by anyone but me). 

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/8adf53ee-8f42-47fb-912e-1d12fe869699)

In situation 1, the colliding faces are aligned so the clipped contact manifold (purple) should be returned. In situation 2, the colliding faces are not aligned and if the distance^2 between the left two points (orange) is bigger than `inSpeculativeContactDistanceSq` the point will be discarded and the contact manifold will only consist of the right point. In the query case you'd only specify a tolerance here, the way it is used in the physics engine is that the speculative contact distance is added to it because it should detect collisions that are further away (the bodies are not touching yet but we still want to know what the manifold will look like if the bodies move far enough in the time step to get into contact).

Reply:
Great, thank you!

> also this function was never intended to be used by anyone but me

I got the sense that this might have been the case, judging by the debug rendering stuff that's in there. Would you rather see that some new (more formal/public) method gets added, or that I go about doing this in some other way, or do I go ahead with this implementation for now?

Reply:
Go ahead for now, it means that it is slightly more likely that I will make a breaking change to your code

Reply:
Will do. That's totally fine.